File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/viewHolders/ShopItemViewHolder.java
package com.habitrpg.android.habitica.ui.viewHolders;

import android.content.Context;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.facebook.drawee.view.SimpleDraweeView;
import com.habitrpg.android.habitica.HabiticaBaseApplication;
import com.habitrpg.android.habitica.R;
import com.habitrpg.android.habitica.models.shops.ShopItem;
import com.habitrpg.android.habitica.ui.helpers.DataBindingUtils;
import com.habitrpg.android.habitica.ui.views.HabiticaIconsHelper;
import com.habitrpg.android.habitica.ui.views.shops.PurchaseDialog;

import butterknife.BindView;
import butterknife.ButterKnife;

public class ShopItemViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

    @BindView(R.id.imageView)
    SimpleDraweeView imageView;
    @BindView(R.id.buyButton)
    View buyButton;
    @BindView(R.id.currency_icon_view)
    ImageView currencyIconView;
    @BindView(R.id.priceLabel)
    TextView priceLabel;

    @BindView(R.id.item_detail_indicator)
    TextView itemDetailIndicator;

    public String shopIdentifier;
    private ShopItem item;

    private Context context;

    Drawable lockedDrawable;
    Drawable limitedDrawable;
    Drawable countDrawable;

    public ShopItemViewHolder(View itemView) {
        super(itemView);

        context = itemView.getContext();

        ButterKnife.bind(this, itemView);

        itemView.setOnClickListener(this);
        itemView.setClickable(true);

        lockedDrawable = new BitmapDrawable(context.getResources(), HabiticaIconsHelper.imageOfItemIndicatorLocked());
        limitedDrawable = new BitmapDrawable(context.getResources(), HabiticaIconsHelper.imageOfItemIndicatorLimited());
        countDrawable = new BitmapDrawable(context.getResources(), HabiticaIconsHelper.imageOfItemIndicatorNumber());
    }

    public void bind(ShopItem item) {
        this.item = item;
        buyButton.setVisibility(View.VISIBLE);

        DataBindingUtils.loadImage(this.imageView, item.getImageName());

        if (item.getUnlockCondition() == null || !item.getLocked()) {
            priceLabel.setText(String.valueOf(item.getValue()));
            if (item.getCurrency().equals("gold")) {
                currencyIconView.setImageBitmap(HabiticaIconsHelper.imageOfGold());
                priceLabel.setTextColor(ContextCompat.getColor(context, R.color.gold));
            } else if (item.getCurrency().equals("gems")) {
                currencyIconView.setImageBitmap(HabiticaIconsHelper.imageOfGem());
                priceLabel.setTextColor(ContextCompat.getColor(context, R.color.green_10));
            } else if (item.getCurrency().equals("hourglasses")) {
                currencyIconView.setImageBitmap(HabiticaIconsHelper.imageOfHourglass());
                priceLabel.setTextColor(ContextCompat.getColor(context, R.color.brand_300));
            } else {
                buyButton.setVisibility(View.GONE);
            }
        } else {
            priceLabel.setText(item.getUnlockCondition().readableUnlockConditionId());
        }

        itemDetailIndicator.setText(null);
        itemDetailIndicator.setVisibility(View.GONE);
        if (item.isLimited()) {
            itemDetailIndicator.setBackground(limitedDrawable);
            itemDetailIndicator.setVisibility(View.VISIBLE);
        }

        if (item.getLocked()) {
            priceLabel.setTextColor(ContextCompat.getColor(context, R.color.gray_300));
            currencyIconView.setAlpha(0.5f);
            itemDetailIndicator.setBackground(lockedDrawable);
            itemDetailIndicator.setVisibility(View.VISIBLE);
        } else {
            currencyIconView.setAlpha(1.0f);
        }
    }

    public void setItemCount(int count) {
        if (count > 0) {
            itemDetailIndicator.setText(String.valueOf(count));
            itemDetailIndicator.setBackground(countDrawable);
            itemDetailIndicator.setVisibility(View.VISIBLE);
        }
    }

    @Override
    public void onClick(View view) {
        PurchaseDialog dialog = new PurchaseDialog(context, HabiticaBaseApplication.getComponent(), item);
        dialog.shopIdentifier = shopIdentifier;
        dialog.show();
    }
}

File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/viewHolders/ShopItemViewHolder.kt
package com.habitrpg.android.habitica.ui.viewHolders

import android.content.Context
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.content.res.AppCompatResources
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import com.habitrpg.android.habitica.R
import com.habitrpg.android.habitica.databinding.RowShopitemBinding
import com.habitrpg.android.habitica.extensions.getImpreciseRemainingString
import com.habitrpg.android.habitica.extensions.getRemainingString
import com.habitrpg.android.habitica.extensions.getShortRemainingString
import com.habitrpg.android.habitica.models.shops.ShopItem
import com.habitrpg.android.habitica.ui.views.HabiticaIconsHelper
import com.habitrpg.common.habitica.extensions.dpToPx
import com.habitrpg.common.habitica.extensions.loadImage

class ShopItemViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
    private val binding = RowShopitemBinding.bind(itemView)
    var shopIdentifier: String? = null
    private var item: ShopItem? = null
    var limitedNumberLeft: Int? = null

    var onNeedsRefresh: (() -> Unit)? = null
    var onShowPurchaseDialog: ((ShopItem, Boolean) -> Unit)? = null

    private var context: Context = itemView.context

    var purchaseCardAction: ((ShopItem) -> Unit)? = null

    var isPinned = false
        set(value) {
            field = value
            binding.pinIndicator.visibility = if (field) View.VISIBLE else View.GONE
        }

    var isCompleted = false
        set(value) {
            field = value
            binding.completedIndicator.visibility = if (field) View.VISIBLE else View.GONE
        }

    init {
        itemView.setOnClickListener(this)
        itemView.isClickable = true
        binding.pinIndicator.setImageBitmap(HabiticaIconsHelper.imageOfPinnedItem())
    }

    fun bind(
        item: ShopItem,
        canBuy: Boolean,
        numberOwned: Int
    ) {
        this.item = item
        binding.buyButton.visibility = View.VISIBLE

        var contentDescription = item.text

        binding.imageView.loadImage(item.imageName?.replace("_locked", ""))

        binding.itemDetailIndicator.text = null
        binding.itemDetailIndicator.visibility = View.GONE

        val lockedReason = item.shortLockedReason(context)
        if (!item.locked || lockedReason == null) {
            binding.priceLabel.text = item.value.toString()
            binding.priceLabel.currency = item.currency
            if (item.currency == null) {
                binding.buyButton.visibility = View.GONE
            }
            binding.priceLabel.visibility = View.VISIBLE
            binding.unlockLabel.visibility = View.GONE
            contentDescription += ", ${item.value} ${binding.priceLabel.currencyContentDescription}"
        } else {
            binding.unlockLabel.text = lockedReason
            binding.priceLabel.visibility = View.GONE
            binding.unlockLabel.visibility = View.VISIBLE
            contentDescription += ", $lockedReason"
        }
        val isLimited = item.isLimited || item.availableUntil != null
        if (numberOwned > 0) {
            binding.itemDetailIndicator.text = numberOwned.toString()
            if (isLimited) {
                binding.itemDetailIndicator.background = AppCompatResources.getDrawable(context, R.drawable.pill_bg_purple_300)
                binding.itemDetailIndicator.setTextColor(ContextCompat.getColor(context, R.color.white))
            } else {
                binding.itemDetailIndicator.background = AppCompatResources.getDrawable(context, R.drawable.pill_bg_gray)
                binding.itemDetailIndicator.setTextColor(ContextCompat.getColor(context, R.color.text_quad))
            }
            binding.itemDetailIndicator.visibility = View.VISIBLE
            contentDescription += ", ${context.getString(R.string.owned)}: $numberOwned"
        } else if (item.locked) {
            binding.itemDetailIndicator.background =
                AppCompatResources.getDrawable(context, if (isLimited) R.drawable.shop_locked_limited else R.drawable.shop_locked)
            binding.itemDetailIndicator.visibility = View.VISIBLE
            contentDescription += ", ${context.getString(R.string.locked)}"
        } else if (isLimited) {
            if (numberOwned == 0) {
                binding.itemDetailIndicator.background = AppCompatResources.getDrawable(context, R.drawable.shop_limited)
            } else {
                binding.itemDetailIndicator.background = AppCompatResources.getDrawable(context, R.drawable.pill_bg_purple_300)
            }
            binding.itemDetailIndicator.visibility = View.VISIBLE

            item.availableUntil?.let {
                contentDescription += ", ${it.getImpreciseRemainingString(context.resources)}"
            }
        }

        val limitedLeft = item.limitedNumberLeft ?: limitedNumberLeft
        if (item.key == "gem" && limitedLeft == -1) {
            binding.itemDetailIndicator.background =
                AppCompatResources.getDrawable(context, R.drawable.item_indicator_subscribe)
            binding.itemDetailIndicator.visibility = View.VISIBLE
            contentDescription += ", ${context.getString(R.string.locked)}"
        } else if (item.key == "gem") {
            binding.itemDetailIndicator.background =
                AppCompatResources.getDrawable(context, R.drawable.pill_bg_green)
            binding.itemDetailIndicator.text = "$limitedLeft"
            binding.itemDetailIndicator.setTextColor(ContextCompat.getColor(context, R.color.white))
            binding.itemDetailIndicator.visibility = View.VISIBLE
            contentDescription += ", ${context.getString(R.string.gems_left_nomax, limitedLeft)}"
        }

        if (binding.itemDetailIndicator.visibility == View.VISIBLE) {
            val layoutParams = binding.itemDetailIndicator.layoutParams
            layoutParams.width =
                if (binding.itemDetailIndicator.text.isBlank()) {
                    24.dpToPx(context)
                } else {
                    ViewGroup.LayoutParams.WRAP_CONTENT
                }
            binding.itemDetailIndicator.layoutParams = layoutParams
        }

        binding.priceLabel.isLocked = item.locked || (!canBuy && item.currency == "gold")

        binding.container.contentDescription = contentDescription
    }

    override fun onClick(view: View) {
        val item = item
        if (item != null && item.isValid) {
            onShowPurchaseDialog?.invoke(item, isPinned)
        }
    }

    fun hidePinIndicator() {
        binding.pinIndicator.visibility = View.GONE
    }
}
